### PR TITLE
Add options to change rendering of colours

### PR DESCRIPTION
**Pull request by EarthenSky**
Just added some final options for changing the method of colouring the output.

https://github.com/Vixlump/GG/assets/24978329/dab1f95d-ed1a-4711-b9b6-2500b04a58d1

https://github.com/Vixlump/GG/assets/24978329/bd031f6b-be40-4acc-a5ab-eabbeb51743e

### CODE REVIEW

File: client/gg_client.cpp [hunk 1]
```diff
@@ -364,7 +387,7 @@ int GG::Client::stream(std::string token) {
 
 	// create colour lookup table for generating ascii colour codes 
 	CharFormat* colour_lookup_table = new CharFormat[16 * 16 * 16];
-    this->_populate_color_lookup_table(colour_lookup_table, ColorSchemeKind::MSE);
+  this->_populate_color_lookup_table(colour_lookup_table, ColorSchemeKind::MSE, NUM_XTERM_COLOURS, false);
```
Comment (micahdbak) [inline client/gg_client.cpp:390]: kewl, so should there be some new "num colors" variable kept in the user session, that could be passed into this function to configure output?

++, how do you configure whether the background is colored or not?

File: client/gg_client.cpp [hunk 2]
```diff
@@ -289,24 +290,46 @@ void GG::Client::_populate_color_lookup_table(CharFormat *color_lookup_table, Co
...
+
+	// fill in holes with closest match
+	if (virtual_color_table.size() < 4096) {
+		for (int r = 0; r < 16; r++) {
+			for (int g = 0; g < 16; g++) {
+				for (int b = 0; b < 16; b++) {
+					int combined_rgb = (r << (16+4)) | (g << (8+4)) | (b << 4);
+					if (virtual_color_table.find(combined_rgb) == virtual_color_table.end()) {
+						CharFormat closest_match;
+						int closest_mse = 256 * 256 * 3;
```
[the remark this answers is not in the document]
Comment (EarthenSky) [reply client/gg_client.cpp:319]: hmmm, not really, without obfuscating it. The three for loops iterate through all 4096 12 bit colours, then combine them into only the 4 MSB of each colour as a single `u32`...

Here, I took the inner part & moved it into a function, which should help a lil.

**Comment by EarthenSky** [comment]
I've noticed a bunch of weird formatting issues, so I wanna see what happens if I run `clang-format`. I'm not sure what the standard formatters are for c++. but it seems nice enough. I'll do another pr for that

File: client/gg_client.cpp (hunk 1 above)
Comment (EarthenSky) [reply client/gg_client.cpp]: yeah! the two parameters we'd want to vary are `num_colours` and `disable_bg_colours`. Both work separately, so if you wanted to stop the background from being coloured, you'd pass true to disable background colours.

**Comment by EarthenSky** [comment]
I'm also gonna use the American color spelling to be consistent with ftxui usage (I tried my best 🥹)

File: client/gg_client.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (micahdbak) [reply client/gg_client.cpp]: Yeah, I assumed as much; I like the current version so thanks!!

File: client/gg_client.cpp (hunk 1 above)
Comment (micahdbak) [reply client/gg_client.cpp]: ohhh I see, cool! This should be easy to embed into the sessions then; thanks!!!

**Comment by micahdbak** [comment]
@EarthenSky , do you think this PR is ready to merge? I'm happy with your changes so far!

**Comment by micahdbak** [comment]
also @Vixlump , wanna review this PR and confirm things work on your end?

**Comment by EarthenSky** [comment]
> do you think this PR is ready to merge?

Yep! It's good now

**Comment by micahdbak** [comment]
teehee. I'll check that it works on my machine and, if yes, I'll merge!

**Comment by micahdbak** [comment]
yup - looks good!

**Pull request merged**